### PR TITLE
Fix race conditions in the concatenate engine streaming

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

While looking at the flaky test pointed out in https://github.com/vitessio/vitess/issues/16641 we found several issues in concatenate engine primitive in the streaming mode. The following are the issues and the fixes in this PR - 

1. In the parallel streaming mode, when we calculate the final field types for the result using`getFieldTypes`, we were overriding the fields of one of the result packets that we got. If this was a packet from a source that needed coercion, because we overwrote the fields, with what we expect them to be, when we check if coercion is required we end up with false even when it requires coercion. This is fixed in the PR by storing the result fields separately and checking for needing coercion with the original fields of the result.
2. Another failure that was found was in the sequential streaming mode. This failure was also masked by the fact that we were not copying the rows when returning the result set in fake primitive. So the only time this test would fail was when ☝️ failed too. If that passed this would pass as well. We have fixed the tests as part of this PR to copy the rows as well before returning them since the concatenate engine modifies them so each test wasn't running with the desired inputs. 
The problem was that we were checking for needing coercion per result packet. But in the streaming mode, if a source is sending multiple packets, then only the first packet will have the fields and the others won't. This leads to the concatenate engine only coercing values in the first packet and not in the subsequent ones since there are no fields. The fix was to check for needing coercion once for a given source and use it for all its packets.
3. Final failure was found after noticing that in the test we weren't checking for the fields in the stream execute output. This has been fixed wherein we now verify we get the correct fields from stream execute as well. Moreover, another check has been added that ensures that all the fields returned as part of the results should match. Previously we would only fix the fields in the first packet that was sent, but the remaining packets could have incorrect names and field types in them because we never overwrote them. The fix is to only send fields for the first result and empty them out for all the others.

## Related Issue(s)

- Fixes https://github.com/vitessio/vitess/issues/16641

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/engine/concatenate.go [hunk 1]
```diff
@@ -408,6 +424,9 @@ func (c *Concatenate) coerceAndVisitResults(
 				}
 			}
 		}
+		if r.Fields != nil {
+			r.Fields = fields
+		}
 		err := callback(r)
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/concatenate.go:430]: nit: this can be set only for the first result set.

Comment (harshit-gangal) [reply go/vt/vtgate/engine/concatenate.go:430]: this can happen outside the for loop.

File: go/vt/vtgate/engine/concatenate.go [hunk 2]
```diff
@@ -245,38 +247,32 @@ func (c *Concatenate) parallelStreamExec(inCtx context.Context, vcursor VCursor,
...
 			for _, row := range res.Rows {
 				if err := c.coerceValuesTo(row, fieldTypes, sqlmode); err != nil {
 					return err
 				}
 			}
 		}
+		if res.Fields != nil {
+			res.Fields = resultFields
+		}
 		return in(res)
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/concatenate.go:276]: This is not needed as we have already passed on the first packet with fields

File: go/vt/vtgate/engine/concatenate.go [hunk 3]
```diff
@@ -296,12 +292,26 @@ func (c *Concatenate) parallelStreamExec(inCtx context.Context, vcursor VCursor,
...
+									srcNeedsCoercion = true
+									break
+								}
+							}
+							needsCoercion[srcIdx] = srcNeedsCoercion
+						}
+
 						muFields.Unlock()
 						defer condFields.Broadcast()
 						return callback(resultChunk, currIndex)
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/concatenate.go:317]: this callback should contain the coerced field type.

File: go/vt/vtgate/engine/concatenate.go (hunk 2 above)
Comment (GuptaManan100) [reply go/vt/vtgate/engine/concatenate.go]: Fixed

File: go/vt/vtgate/engine/concatenate.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/vtgate/engine/concatenate.go]: Fixed

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**